src/config.ts

```diff
@@ -1,361 +1,153 @@
 }
 
-/** Updates the config to the latest version.
-    Proposed semantic for config versionning:
+/** Updates the config to the latest version. Proposed semantic for config
+    versionning:
      - x.y -> x+1.0 : major architectural changes
-     - x.y -> x.y+1 : renaming settings/changing their types
-    There's no need for an updater if you're only adding a new setting/changing
-    a default setting
-
-    When adding updaters, don't forget to set("configversion", newversionnumber)!
+     - x.y -> x.y+1 : renaming settings/changing their types There's no need for
+       an updater if you're only adding a new setting/changing a default setting
 */
-export async function update() {
-    let updaters = {
-        "0.0": async () => {
-            try {
-                // Before we had a config system, we had nmaps, and we put them in the
-                // root namespace because we were young and bold.
-                let legacy_nmaps = await browser.storage.sync.get("nmaps")
-                if (legacy_nmaps) {
-                    USERCONFIG["nmaps"] = Object.assign(legacy_nmaps["nmaps"], USERCONFIG["nmaps"])
-                }
-            } finally {
-                set("configversion", "1.0")
-            }
-        },
-        "1.0": () => {
-            let vimiumgi = getDeepProperty(USERCONFIG, "vimium-gi")
-            if (vimiumgi === true || vimiumgi === "true")
-                set("gimode", "nextinput")
-            else if (vimiumgi === false || vimiumgi === "false")
-                set("gimode", "firefox")
-            unset("vimium-gi")
-            set("configversion", "1.1")
-        }
+export async function migrate(cfgObj: object): Promise<object> {
+    let ver = getDeepProperty(cfgObj, ['configversion'])
+    if (!ver) ver = '0.0'
+
+    // Use fall-through to migrate to latest version
+    switch (ver) {
+        case '0.0':
+            await migrate_00_to_10(cfgObj)
+        case '1.0':
+            migrate_10_to_11(cfgObj)
     }
-    if (!get("configversion"))
-        set("configversion", "0.0")
-    while (updaters[get("configversion")] instanceof Function) {
-        await updaters[get("configversion")]()
-    }
-}
 
-/** Read all user configuration from storage API then notify any waiting asynchronous calls
+    setDeepProperty(cfgObj, '1.1', ['configversion'])
+    return cfgObj
+
+    async function migrate_00_to_10(cfgObj: object): Promise<object> {
+        try {
+            // Before we had a config system, we had nmaps, and we put them in the
+            // root namespace because we were young and bold.
+            let legacy_nmaps = await browser.storage.sync.get('nmaps')
+            if (legacy_nmaps)
+                cfgObj['nmaps'] = Object.assign(
+                    legacy_nmaps['nmaps'],
+                    cfgObj['nmaps']
+                )
+        } catch (e) {
+            console.error('Error while migration from 0.0 -> 1.0', e)
```

Comment:
I get this error with the message "can't convert undefined to object" on every page that loads (on a fresh profile - i.e. starting config version is 1.1).